### PR TITLE
Merge release/v3.0 into release/v4.0

### PR DESCRIPTION
**Pull request by mallardduck**
This should be on hold and then rebased to merge after:

- #429 
- #430

### CODE REVIEW

**Comment by mallardduck** [comment]
🤔 - I intentionally did this as a squashed commit locally, wasn't expecting this many commits when I pushed and created PR. Either way as mentioned in description this should be on hold until those others merge then i'll rebase at that point in time and address this oddness.

**Review by ericpromislow — APPROVED** [review]
LGTM

**Pull request merged**
